### ISSUE TITLE
mutation operators

### ISSUE THREAD

**Issue by donghwan-shin** [yinahou/CAIN-code#1]
Hi Ziyu, thanks for making the repository. 

I saw the mutation generation code, and below are my comments

### AOR (Arithmetic Operator Replacement)

This looks good to me.

### LOR (Logical operator replacement)

This looks okay, but don't we need to consider `|` and `&` in addition to `||` and `&&`? 

### COR (Conditional Operator replacement)

This operator can be implemented to generate the following mutants:
- `if(e)` -> `if(true)`
- `if(e)` -> `if(false)`

However, this seems too easy to detect from LLMs. Let's just remove this operator from our code since we already found that this does not apply to some languages. 

### ROR (Relational operator replacement)

It should be updated as follows:

```python
        for m in re.finditer('<', original_sol):
            mutations.append(replace_once(original_sol, '<', '>', m))
            mutations.append(replace_once(original_sol, '<', '<=', m))
            mutations.append(replace_once(original_sol, '<', '>=', m))
            mutations.append(replace_once(original_sol, '<, '==', m))
            mutations.append(replace_once(original_sol, '<', '!=', m))
```

It's okay to keep `==` and `!=` to change only for the opposite to avoid unnecessary compiler errors. But the other operators (<, >, <=, >=) should be changed to all possible relation operators as above. 

### STD (Statement deletion)

Your implementation looks okay to me. It might delete some important variable declaration lines, leading to a compiler error, but this type of change should be detected by LLMs properly. So, let's keep this.

### LVR (Literal Value Replacement)

This looks good.

### ABS (Absolute value insertion)

Ideally, ABS should be applied to all expressions `e` if the type of `e` is numeric (e.g., int, float, double). 
Also, it should generate the following three mutants according to Table 1 [1]: 
- `e` -> 0
- `e` -> `abs(e)`
- `e` -> `-abs(e)`

The current implementation (copied below) looks incomplete since (1) it simply applies `abs()` to any expression regardless of the type, and (2) it does not generate `0` or `-abs(e)`. Can you fix this?

https://github.com/yinahou/CAIN-code/blob/1d1e8dd791c731f8b3fdb8803652be1943828847/mutation_generation.ipynb#L114-L119

### UOI (Unary operator insertion)

It is supposed to generate the following mutants:
- `e` -> `-e`
- `e` -> `--e`
- `e` -> `++e`
- `e` -> `not e`

Some of the above might not apply to some programming languages (I am unsure). Can you please have a look at this again?

## Summary

To sum up, these are my comments:
- AOR: Good
- LOR: Need improvement
- COR: Remove this
- ROR: Need improvement
- STD: Good
- LVR: Good
- ABS: Need improvement (if difficult, we can remove this since this looks easy to detect)
- UOI: Need improvement (if difficult, we can remove this since this looks easy to detect)

In our paper, we can justify the inclusion of mutation operators as follows:
(1) widely used in the existing literature, 
(2) applicable to all programming languages.
And then, we can say, "using more (specific) mutation operators for each programming language is possible but not in the scope of this work". 

### References
1. Papadakis, Mike, et al. "Mutation testing advances: an analysis and survey." Advances in Computers. Vol. 112. Elsevier, 2019. 275-378.
